docs/docs/state-and-lifecycle.md

```diff
@@ -0,0 +1,504 @@
+
+State is similar to props, but it is private and fully controlled by the component.
+
+We [mentioned before](/react/docs/components-and-props.html#functional-and-class-components) that components defined as classes have some additional features. Local state is exactly that: a feature available only to classes.
+
+## Converting a Function to a Class
+
+You can convert a functional component like `Clock` to a class in five steps.
+
+1) Create an [ES2015 class](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes) with the same name that extends `React.Component`:
+
+```js{1-3}
+class Clock extends React.Component {
+
+}
+
+function Clock(props) {
+  return (
+    <div>
+      <h1>Hello, world!</h1>
+      <h2>It is {props.date.toLocaleTimeString()}.</h2>
+    </div>
+  );
+}
+```
+
+2) Add a single empty method to it called `render()`:
+
+```js{2-4}
+class Clock extends React.Component {
+  render() {
+
+  }
+}
+
+function Clock(props) {
+  return (
+    <div>
+      <h1>Hello, world!</h1>
+      <h2>It is {props.date.toLocaleTimeString()}.</h2>
+    </div>
+  );
+}
+```
+
+3) Move the body of the function into the `render()` method:
+
+```js{3-8}
+class Clock extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Hello, world!</h1>
+        <h2>It is {props.date.toLocaleTimeString()}.</h2>
+      </div>
+    );
+  }
+}
+
+function Clock(props) {
+
+}
+```
+
+4) Replace `props` with `this.props` in the `render()` body:
+
+```js{6}
+class Clock extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Hello, world!</h1>
+        <h2>It is {this.props.date.toLocaleTimeString()}.</h2>
+      </div>
+    );
+  }
+}
+
+function Clock(props) {
+
+}
+```
+
+5) Delete the empty function declaration:
+
+```js
+class Clock extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Hello, world!</h1>
+        <h2>It is {this.props.date.toLocaleTimeString()}.</h2>
+      </div>
+    );
+  }
+}
+```
+
+[Try it on Codepen.](http://codepen.io/gaearon/pen/zKRGpo?editors=0010)
+
+`Clock` is now defined as a class rather than a function.
+
+This lets is use additional features such as local state and lifecycle hooks. We will show how to use them to make `Clock` reusable and encapsulated below.
+
+## Adding Local State to a Class
+
+We will move the `date` from props to state in three steps:
+
+1) Replace `this.props.date` with `this.state.date` in the `render()` method:
+
+```js{6}
+class Clock extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Hello, world!</h1>
+        <h2>It is {this.state.date.toLocaleTimeString()}.</h2>
+      </div>
+    );
+  }
+}
+```
+
+2) Add a [class constructor](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes#Constructor) that assigns the initial `this.state`:
+
+```js{2-5}
+class Clock extends React.Component {
+  constructor(props) {
+    super(props);
```

Comment:
@azizj1 The pitfall is that `this.props` will be undefined inside constructor if you don't pass `props` to `super`. This can be confusing so we recommend always passing them for consistency.